### PR TITLE
Crewmates can do tasks!

### PR DESCRIPTION
**Pull request by rictic**
Get close to 'em and press E! Boom, done! Everone else will know, and no one will care!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -418,6 +369,74 @@ impl Game {
...
+            None => return Ok(()), // not controlling anything
+            Some(c) => *c,
+        };
+
+        let mut killed_player: Option<&mut Player> = None;
+        let mut closest_distance = self.kill_distance;
+
+        for player in self.players.iter_mut() {
+            // TODO(can't kill impostors)
+            if player.color == local_player_color || player.dead {
```
Comment (cdata) [inline src/lib.rs:383]: I wonder if players could support an `is_local` field 🤔

File: src/lib.rs [hunk 2]
```diff
@@ -418,6 +369,74 @@ impl Game {
...
+            if let Some(player) = self.local_player_mut() {
+                player.position = body.position;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn activate_near(&mut self, position: Position) -> Result<(), JsValue> {
+        let mut closest_distance = self.kill_distance;
```
Comment (cdata) [inline src/lib.rs]: Should there be a distinct `activation_distance`?

File: src/lib.rs [hunk 3]
```diff
@@ -418,6 +369,74 @@ impl Game {
...
+        let local_player = match self.local_player_mut() {
+            Some(player) => player,
+            None => return Ok(()),
+        };
+        let is_imp = local_player.impostor;
+        let color = local_player.color;
+
+        let mut finished_task: Option<(usize, &mut Task)> = None;
+        for (i, task) in local_player.tasks.iter_mut().enumerate() {
```
Comment (cdata) [inline src/lib.rs:422]: I love how intuitive it is to read this form of iteration even though I have never read or written it myself.

File: src/lib.rs [hunk 4]
```diff
@@ -418,6 +369,74 @@ impl Game {
...
+            // TODO(can't kill impostors)
+
+            let distance = position.distance(task.position);
+            if distance < closest_distance {
+                finished_task = Some((i, task));
+                closest_distance = distance;
+            }
+        }
+        if let Some((i, task)) = finished_task {
```
Comment (cdata) [inline src/lib.rs:431]: It took my brain a second to accept that this was intentional assignment. Something about the `Some` being on the LHS disrupted my expectations...

File: src/lib.rs [hunk 5]
```diff
@@ -443,6 +462,15 @@ impl Game {
                     }
                 }
             }
+            Message::FinishedTask(FinishedTask { color, index }) => {
```
Comment (cdata) [inline src/lib.rs:465]: I don't quite get what is going on here. It looks like a call (`Message::FinishedTask(...)`) but then has a fat arrow and a block with some stuff inside. As a guess, I might think this performs some network side-effect and then calls back and does whatever work is in the block. But, that feels like a wild guess. I could probably grok it better if I read wherever `FinishedTask` is defined..

Comment (cdata) [reply src/lib.rs:465]: Okay, I read that `Message::FinishedTask` is an enum, so actually this looks more like a call signature for an anonymous function. And, unfurling the code above, I now see we are in a `match` clause. So, I get it 😅 I think...

File: www/index.js [hunk 6]
```diff
@@ -65,16 +65,20 @@ function average(arr) {
...
 const heldButtons = {};
 function updateInput() {
   const up = heldButtons['w'] || heldButtons['arrowup'];
   const down = heldButtons['s'] || heldButtons['arrowdown'];
   const left = heldButtons['a'] || heldButtons['arrowleft'];
   const right = heldButtons['d'] || heldButtons['arrowright'];
-  const q = heldButtons['q'];
-  game.set_inputs(up, down, left, right, q);
+  const kill = heldButtons['q'];
+  const report = heldButtons['r'];
```
Comment (cdata) [inline www/index.js:79]: 🚨📣

**Review by cdata — APPROVED** [review]

File: src/lib.rs (hunk 4 above)
Comment (rictic) [reply src/lib.rs:431]: Yeah, `if let ...` pattern matching takes some getting used to. I'm currently in favor of it, but I'm definitely doing more writing than reading at the moment

File: src/lib.rs (hunk 3 above)
Comment (rictic) [reply src/lib.rs:422]: it's a little odd to me that you've gotta do `.iter()` and `.iter_mut()` but it's growing on me.

File: src/lib.rs (hunk 5 above)
Comment (rictic) [reply src/lib.rs:465]: Yeah, you got it, this is a pattern to match that also pulls `color` and `index` into local variables in the case where the incoming message is a FinishedTask message

File: src/lib.rs (hunk 1 above)
Comment (rictic) [reply src/lib.rs:383]: In a PR a few down the line I'm switching to UUIDs for identifying players

File: src/lib.rs (hunk 2 above)
Comment (rictic) [reply src/lib.rs]: +1, done

**Pull request merged**